### PR TITLE
Improve circularQueue and fixed cleanup().

### LINKED ISSUE #75: CircularQueue should not exceed its capacity.

### PR DESCRIPTION
**Pull request by mond77**
close #75 

changed:
1. fix that cleanup() sometimes doesn't work.
2. when calling client::new() a daemon thread will be created for cleanup loop. (that avoid concurrency raced so atomicU32 could replace mutex)
3. add mpsc::sync_channel(1) named `occupied `; block register_callback if the id has already been occupied.

tested: 
10000 Tokio task for call_remote() every 20 seconds, totally 10 times. 

result:
when main exited, start_index was 78192.(forget to wait all callbacks done)
cleanup can reasonably work.
call_remote without occurring error.

### CODE REVIEW

File: src/rpc/connection.rs [hunk 1]
```diff
@@ -510,21 +512,22 @@ impl OperationCallback {
...
             channel: mpsc::channel(),
+            occupied: mpsc::sync_channel(1),
         }
     }
 }
 
 pub struct CircularQueue {
     callbacks: Vec<*const OperationCallback>,
-    start_index: Arc<Mutex<u32>>, // maybe we can use a lock-free queue
+    start_index: Arc<AtomicU32>, // maybe we can use a lock-free queue
```
Comment (luanshaotong) [inline src/rpc/connection.rs:522]: Pls remove the out-of-date comment.

File: src/rpc/connection.rs [hunk 2]
```diff
@@ -554,40 +557,21 @@ impl CircularQueue {
         rsp_meta_data: &mut [u8],
         rsp_data: &mut [u8],
     ) -> Result<u32, Box<dyn std::error::Error>> {
-        let id = self.end_index.fetch_add(1, Ordering::Relaxed);
+        let idx = self.end_index.fetch_add(1, Ordering::Relaxed);
+        let id = idx % REQUEST_QUEUE_LENGTH as u32;
```
Comment (luanshaotong) [inline src/rpc/connection.rs:561]: Pls make sure that integer overflow issues do not affect.

**Comment by luanshaotong** [comment]
Another problem is that the new channel may increase latency. Pls test it in the future.

File: src/rpc/connection.rs (hunk 2 above)
Comment (mond77) [reply src/rpc/connection.rs:561]: change the mod length will be ok.
Screen Shot 2023-01-30 at 12 38 56

File: src/rpc/connection.rs [hunk 3]
```diff
@@ -685,13 +667,48 @@ impl CircularQueue {
...
+                    CallbackState::Empty => {
+                        debug!("unexpected behavior happen")
+                    }
+                }
+                if let Ok(()) = (*(callback as *mut OperationCallback)).occupied.1.recv() {
+                    // match Ok(())
+                    debug!("index {idx} has already been cleaned up. ");
+                }
+            }
+            idx += 1;
```
Comment (mond77) [inline src/rpc/connection.rs:698]: 这里idx是u32类型
debug模式下会检查u32是否溢出，溢出会panic；release模式则不会；AtomicU32两种模式都不会

File: src/rpc/connection.rs [hunk 4]
```diff
@@ -685,13 +667,48 @@ impl CircularQueue {
     }
 }
 
+pub fn clean_up(queue: Arc<CircularQueue>) {
+    debug!("start to cleanup");
+    loop {
+        let mut idx = queue.start_index.load(Ordering::Acquire);
+        let end_flag = queue.end_index.load(Ordering::Acquire);
```
Comment (luanshaotong) [inline src/rpc/connection.rs:674]: Mod `idx` and `end_flag` by REQUEST_QUEUE_LENGTH here may help.

File: src/rpc/connection.rs (hunk 2 above)
Comment (luanshaotong) [reply src/rpc/connection.rs:561]: Great!

File: src/rpc/connection.rs (hunk 4 above)
Comment (mond77) [reply src/rpc/connection.rs:674]: there is no need to mod.
the sub `end_flag - idx` contains information of how many callbacks are working.
If the sub is multiple times of REQUEST_QUEUE_LENGTH, we could know how many callbacks are blocking.

File: src/rpc/connection.rs [hunk 5]
```diff
@@ -685,13 +667,48 @@ impl CircularQueue {
...
+        while idx != end_flag {
```
Comment (mond77) [inline src/rpc/connection.rs:675]: here the condition is changed so idx could overflow to catch up end_flag

**Review by luanshaotong — APPROVED** [review]
LGTM

**Pull request merged**